opentelemetry-sdk/src/logs/growable_array.rs

```diff
@@ -0,0 +1,323 @@
+use std::array::IntoIter;
+
+/// The default max capacity for the stack portion of `GrowableArray`.
+const DEFAULT_MAX_STACK_CAPACITY: usize = 10;
+/// The default initial capacity for the vector portion of `GrowableArray`.
+const DEFAULT_INITIAL_VEC_CAPACITY: usize = 5;
+
+#[derive(Debug, Clone, PartialEq)]
+/// A hybrid vector that starts with a fixed-size array and grows dynamically with a vector.
+///
+/// `GrowableArray` uses an internal fixed-size array (`initial`) for storing elements until it reaches
+/// `MAX_STACK_CAPACITY`. When this capacity is exceeded, additional elements are stored in a heap-allocated
+/// vector (`additional`). This structure allows for efficient use of stack memory for small numbers of elements,
+/// while still supporting dynamic growth.
+///
+pub(crate) struct GrowableArray<
+    T: Default + Clone + PartialEq,
+    const MAX_STACK_CAPACITY: usize = DEFAULT_MAX_STACK_CAPACITY,
```

Comment:
I would be in favor of removing these `const` generics for simplicity.

Reply:
I tried the rust ideomatic way of specifing the size as compile time constants, taking _inspiration_ from rust-lang array, and `smallvec`. But happy to change if this looks complex. What do you suggest?

Reply:
Since this we are not marking this data structure public and our usage of it doesn't involve dynamic configuration, I think we should keep the code simple and less for easier maintenance.

Reply:
Are you suggesting replacing the `const` generics with the `const` variable?

```rust
const MAX_STACK_CAPACITY = 5;
const INITIAL_VEC_CAPACITY =10;
pub(crate) struct GrowableArray
{
    initial: [T; MAX_STACK_CAPACITY ],
    additional: Option<Vec<T>>,
    count: usize,
}
```

My initial thoughts were that this would allow flexibility if we want to introduce the feature-flag, for heap-only allocation requirements:
```rust
#[cfg(feature = "heap-only")]
pub(crate) type AttributesGrowableArray =
    GrowableArray<Option<(Key, AnyValue)>, 0, 10>;
#[cfg(not(feature = "heap-only"))]
pub(crate) type AttributesGrowableArray =
    GrowableArray<Option<(Key, AnyValue)>, 10>;
```

Reply:
> Are you suggesting replacing the const generics with the const variable?

Yes.

> My initial thoughts were that this would allow flexibility if we want to introduce the feature-flag, for (say) heap-only allocation requirements:

Is there any demand for heap-only allocation? Even if it's there, that should still be possible without using `const` generics, I think. We could simply update the `const` value. Would something like this work?
```rust
const MAX_STACK_CAPACITY: usize = match () {
    #[cfg(feature = "heap-only")]
    () => 0,
    _ => 10
  };

const INITIAL_VEC_CAPACITY : usize = match () {
    #[cfg(feature = "heap-only")]
    () => 10,
    _ => 0,
  };
```

Reply:
> Would something like this work?

This will work, but then we need to add more such constants if we want to later use this array in trace/event attributes with different size/feature requirements. The existing approach provides a cleaner way to specify these constants at the place of creation. I do agree `const` variable is a much simpler approach if we don't plan to use this array at other places. We can revisit it later and change it according?